core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -412,12 +416,69 @@ public final SegmentsStats segmentsStats() {
+                memoryResources.addStoredFieldsInBytes(guardedRamBytesUsed(segmentReader.getFieldsReader()));
+                memoryResources.addTermVectorsInBytes(guardedRamBytesUsed(segmentReader.getTermVectorsReader()));
+                memoryResources.addNormsInBytes(guardedRamBytesUsed(segmentReader.getNormsReader()));
+                memoryResources.addDocValuesInBytes(guardedRamBytesUsed(segmentReader.getDocValuesReader()));
+                stats.addMemoryIndexResources(memoryResources);
+                
+                Directory directory = null;
+                SegmentInfo segmentInfo = segmentReader.getSegmentInfo().info;
+                boolean useCompoundFile = segmentInfo.getUseCompoundFile();
+                if (useCompoundFile) {
+                    try {
+                        directory = engineConfig.getCodec().compoundFormat().getCompoundReader(segmentReader.directory(), segmentInfo, IOContext.READ);
+                    } catch (IOException e) {
+                        logger.trace("error when opening compound reader for segment [{}]", e, segmentInfo.name);
+                    }
+                } else {
+                    directory = segmentReader.directory();
+                }
+                
+                // Couldn't open compound reader, skip
+                if (directory == null) {
+                    continue;
+                }
+                
+                try {
+                    String[] files = directory.listAll();
+                    for (String file : files) {
+                        String extension = IndexFileNames.getExtension(file);
+                        if (extension == null) {
+                            continue;
+                        }
+                        
+                        long length = directory.fileLength(file);
+                        
+                        if ("tip".equals(extension) || "tim".equals(extension)
+                                || "pos".equals(extension) || "doc".equals(extension) || "pay".equals(extension)) {
+                            diskResources.addTermsInBytes(length);
+                        } else if ("fdt".equals(extension) || "fdx".equals(extension)) {
+                            diskResources.addStoredFieldsInBytes(length);
+                        } else if ("dvd".equals(extension) || "dvm".equals(extension)) {
+                            diskResources.addDocValuesInBytes(length);
+                        } else if ("nvd".equals(extension) || "nvm".equals(extension)) {
+                            diskResources.addNormsInBytes(length);
+                        } else if ("tvd".equals(extension) || "tvx".equals(extension) || "tvf".equals(extension)) {
+                            diskResources.addTermVectorsInBytes(length);
+                        }
+                    }
+                } catch (IOException e) {
+                    // TODO maybe catch specific exceptions for gone files?
```

Comment:
Actually, I think IW should not remove files that are being used by NRT readers (we fixed this a while back in Lucene), so I think we should continue to log the exception, but can you increase log level to WARN?

Also,  can you shrink wrap the exception handling?  Put one try/finally around the `listAll`, and another on the `directory.fileLength`?  And include in the exception message which directory (`directory.toString()`) and which file caused an exception.